File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/intake/Pendulum.java
package org.firstinspires.ftc.teamcode.intake;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.Servo;
import org.firstinspires.ftc.teamcode.utils.systems.ManualMechanismTeleOp;
import org.firstinspires.ftc.teamcode.utils.systems.ServoPositionMechanism;

import static org.firstinspires.ftc.teamcode.Positions.Intake.Pendulum.pickup;
import static org.firstinspires.ftc.teamcode.Positions.Intake.Pendulum.pickupWait;
import static org.firstinspires.ftc.teamcode.Positions.Intake.Pendulum.transfer;

/**
 * Class representing the vertical rotation of the intake.
 * It extends the ServoPositionMechanism class to control the rotation servo.
 */
public class Pendulum extends ServoPositionMechanism {
    private final HardwareMap hardwareMap;
    /**
     * @param hardwareMap the hardware map to get the servo from.
     */
    public Pendulum(HardwareMap hardwareMap) {
        super(transfer);
        this.hardwareMap = hardwareMap;
    }
    /**
     * Gets the servo associated with the Rotate mechanism.
     * @return an array containing the rotation servo.
     */
    @Override
    protected Servo[] getServos() {
        return new Servo[]{hardwareMap.get(Servo.class, "Intake Pendulum")};
    }
}

/**
 * TeleOp class for testing the Pendulum mechanism manually.
 */
@TeleOp(name = "Intake Pendulum Test", group = "C Intake")
class PendulumTest extends ManualMechanismTeleOp {
    public PendulumTest() {
        super(Pendulum::new);
    }
}

/**
 * TeleOp class for testing the Pendulum mechanism positions.
 * @noinspection DuplicatedCode
 */
@TeleOp(name = "Intake Pendulum Positions Test", group = "D Intake")
@Config
class PendulumPositions extends LinearOpMode {
    public static double current = 0.0;
    @Override
    public void runOpMode() {
        Pendulum pendulum = new Pendulum(hardwareMap);
        waitForStart();
        while (opModeIsActive()) {
            if (gamepad1.dpad_right) current = pickupWait;
            else if (gamepad1.dpad_down) current = pickup;
            else if (gamepad1.dpad_up) current = transfer;
            else current += pendulum.getMultiplier() * (gamepad1.right_trigger - gamepad1.left_trigger);
            pendulum.setTargetPosition(current);

            TelemetryPacket packet = new TelemetryPacket();
            if(!pendulum.run(packet)) requestOpModeStop();
            FtcDashboard.getInstance().sendTelemetryPacket(packet);
            telemetry.addData("Current Position", current);
            telemetry.update();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/intake/Pendulum.kt
package org.firstinspires.ftc.intake

import com.acmerobotics.dashboard.config.Config
import com.qualcomm.robotcore.eventloop.opmode.Disabled
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.Servo
import org.firstinspires.ftc.intake.Pendulum.Companion.pickup
import org.firstinspires.ftc.intake.Pendulum.Companion.pickupWait
import org.firstinspires.ftc.intake.Pendulum.Companion.transfer
import org.firstinspires.ftc.utils.PressAction
import org.firstinspires.ftc.utils.systems.ManualMechanismTeleOp
import org.firstinspires.ftc.utils.systems.Movement
import org.firstinspires.ftc.utils.systems.ServoPositionMechanism

@Config
class Pendulum(hardwareMap: HardwareMap) : ServoPositionMechanism(transfer) {
    companion object {
        /**
         * The position of the intake pendulum when it is outside of the submersible.
         */
        @JvmField
        @Volatile
        var pickupWait = 0.1
        /**
         * The position of the intake pendulum when it is inside of the submersible.
         */
        @JvmField
        @Volatile
        var pickup = 0.05
        /**
         * The position of the intake pendulum when it is in the transfer position.
         */
        @JvmField
        @Volatile
        var transfer = 0.75
    }
    /**
     * Gets the servo associated with the Pendulum mechanism.
     */
    override val servos: Array<Servo> = arrayOf(hardwareMap.servo["Intake Pendulum"])
}
/**
 * TeleOp class for testing the Pendulum mechanism manually.
 */
@TeleOp(name = "Intake Pendulum Test", group = "C")
@Disabled
private class PendulumTest : ManualMechanismTeleOp(::Pendulum)
/**
 * TeleOp class for testing the positions of the Pendulum mechanism.
 */
private class PendulumPositions : Movement() {
    /**
     * The Pendulum mechanism to be tested.
     */
    override lateinit var system: Pendulum

    override fun systemInit() {
        system = Pendulum(hardwareMap)

        actions.add(PressAction(gamepad1::dpad_right) { system.targetPosition = pickupWait })
        actions.add(PressAction(gamepad1::dpad_down) { system.targetPosition = pickup })
        actions.add(PressAction(gamepad1::dpad_up) { system.targetPosition = transfer })
    }

    override fun systemLoop() {
        // No need for a loop
    }
}
